frontend/model/contracts/group.js

```diff
@@ -232,6 +232,158 @@ DefineContract({
+          for (const toUser in paymentsFrom[fromUser]) {
+            if (toUser === getters.ourUsername) {
+              for (const paymentHash of paymentsFrom[fromUser][toUser]) {
+                const { data, meta } = allPayments[paymentHash]
+
+                payments.push({ hash: paymentHash, data, meta })
+              }
+            }
+          }
+        }
+
+        return payments
+      })()
+      const todo = (() => {
+        const payments = []
+        const unadjusted = getters.groupIncomeDistribution.filter(p => p.from === ourUsername)
+
+        for (const p of getters.groupIncomeAdjustedDistribution) {
+          if (p.from !== ourUsername) { continue }
+          const existPayment = unadjusted.find(({ to }) => to === p.to) || { amount: 0 }
+          const amount = +currency.displayWithoutCurrency(p.amount)
+          const existingAmount = +currency.displayWithoutCurrency(existPayment.amount)
+
+          if (amount <= 0) { continue }
+
+          const partialAmount = existingAmount - amount
+          const existingPayment = {}
+          if (partialAmount > 0) {
+            // TODO/BUG this only work if the payment is done in 2 parts. if done in >=3 won't work.
+            const sentPartial = sent.find((s) => s.username === p.to && s.amount === partialAmount)
+            if (sentPartial) {
+              existingPayment.hash = sentPartial.hash // Q: Why do we need this hash?
+            }
+          }
+
+          payments.push({
+            ...existingPayment,
+            ...p,
+            total: existingAmount,
+            partial: partialAmount > 0
+          })
+        }
+        return payments
+      })()
+      const toBeReceived = (() => {
+        const unadjusted = getters.groupIncomeDistribution.filter(p => p.to === ourUsername)
+        const payments = []
+        for (const p of getters.groupIncomeAdjustedDistribution) {
+          if (p.to !== ourUsername) { continue }
+
+          const existPayment = unadjusted.find(({ from }) => from === p.from) || { amount: 0 }
+          const amount = +currency.displayWithoutCurrency(p.amount)
+          const existingAmount = +currency.displayWithoutCurrency(existPayment.amount)
+
+          if (amount <= 0) { continue }
+
+          const partialAmount = existingAmount - amount
+          const existingPayment = {}
+          if (partialAmount > 0) {
+            // TODO/BUG this only work if the payment is done in 2 parts. if done in >=3 won't work.
+            const receivePartial = received.find((r) => r.username === p.to && r.amount === partialAmount)
+            if (receivePartial) {
+              existingPayment.hash = receivePartial.hash
+            }
+          }
+          payments.push({
+            ...existingPayment,
+            ...p,
+            total: existingAmount,
+            partial: partialAmount > 0
+          })
+        }
+        return payments
+      })()
+
+      return { sent, todo, received, toBeReceived }
+    },
```

Comment:
@sandrina-p Can you simplify and dry this entire function? It is huge. Right now it is hard for me to follow and read. It contains a billion loops. Can you please:

1. Reduce the number of loops that occur in this function? This means: using `for () { }` instead of higher-order-functions like `.map().find().filter().push()` <- each one of those function calls is a loop. You can replace them with a single loop instead, for example, look at how `paymentsTodo()` is implemented in `Payments.vue`
2. Call getters only once in the entire function and cache any results. For example, there are two calls to `getters.groupIncomeDistribution` in this function. There should be only one.

Using these types of techniques, see if you can reduce the size of this function by at least 50%